[BUG] version check SEGV in toolexec mode on go 1.23

<!--
If you have identified a specific bug in the orchestrion codebase, please describe it here. If the
issue requires debugging or investigation of your specific private setup, that is best accomplished via support. Please contact Datadog [support](http://docs.datadoghq.com/help/).
-->

**Version of orchestrion**

Latest & main

**Describe what happened:**

The call `tools/packages.Load` will return with no errors with `pkg.Module` being `nil`. This being not checked create this error:

```
❯ go test -v -tags=integration,buildtag -shuffle=on -toolexec "$PWD/../orchestrion toolexec" ./tests/redigo
# github.com/docker/docker/errdefs
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0xdeaed6]

goroutine 1 [running]:
github.com/DataDog/orchestrion/internal/ensure.goModVersion({0x0?, 0xc00023bc38?})
        /home/eliott/dd/orchestrion/internal/ensure/requiredversion.go:150 +0x116
github.com/DataDog/orchestrion/internal/ensure.requiredVersion(0x10fb8bf?, 0x11548e0, 0x1155108, {0xc0001b8000, 0xe, 0xe})
        /home/eliott/dd/orchestrion/internal/ensure/requiredversion.go:66 +0x6a
github.com/DataDog/orchestrion/internal/ensure.RequiredVersion(...)
        /home/eliott/dd/orchestrion/internal/ensure/requiredversion.go:44
github.com/DataDog/orchestrion/internal/pin.init.0()
        /home/eliott/dd/orchestrion/internal/pin/pin.go:146 +0x68
```

**Describe what you expected:**

I guess we expect orchestrion execution to not happen in this case because we have no way to make sure we are running the right version of orchestrion.

**Steps to reproduce the issue:**

This issue only happen on certain conditions:
* Need at least go 1.23
* Need to use orchestrion in TOOLEXEC mode (meaning with `-toolexec`, not with `orchestrion go ...`)
* Only some packages of the `github.com/docker/docker` repository are reproductiing this issue

Then running any integration test trying to build `github.com/docker/docker` reproduce the issue

--- a/internal/ensure/requiredversion.go
+++ b/internal/ensure/requiredversion.go
@@ -147,6 +147,7 @@ func goModVersion(dir string) (moduleVersion string, moduleDir string, err error
 		return "", "", errors.Join(errs...)
 	}
 
+	// Shouldn't happen but does: see https://github.com/golang/go/issues/65816
 	if pkg.Module == nil {
 		return "", "", fmt.Errorf("no module information found for package %q", pkg.PkgPath)
 	}

--- a/internal/ensure/requiredversion.go
+++ b/internal/ensure/requiredversion.go
@@ -147,6 +147,10 @@ func goModVersion(dir string) (moduleVersion string, moduleDir string, err error
 		return "", "", errors.Join(errs...)
 	}
 
+	if pkg.Module == nil {
+		return "", "", fmt.Errorf("no module information found for package %q", pkg.PkgPath)
+	}
+
 	if pkg.Module.Replace != nil {
 		// If there's a replace directive, that's what we need to be honoring instead.
 		return pkg.Module.Replace.Version, pkg.Module.Replace.Dir, nil

--- a/internal/ensure/requiredversion.go
+++ b/internal/ensure/requiredversion.go
@@ -147,7 +147,9 @@ func goModVersion(dir string) (moduleVersion string, moduleDir string, err error
 		return "", "", errors.Join(errs...)
 	}
 
-	// Shouldn't happen but does: see https://github.com/golang/go/issues/65816
+	// Shouldn't happen but does when the current working directory is not
+	// part of a go module's source tree.
+	// See: https://github.com/golang/go/issues/65816
 	if pkg.Module == nil {
 		return "", "", fmt.Errorf("no module information found for package %q", pkg.PkgPath)
 	}

--- a/internal/ensure/requiredversion_test.go
+++ b/internal/ensure/requiredversion_test.go
@@ -85,6 +85,23 @@ func TestGoModVersion(t *testing.T) {
 			}
 		})
 	}
+
+	t.Run("no-go-mod", func(t *testing.T) {
+		tmp, err := os.MkdirTemp("", "ensure-*")
+		require.NoError(t, err, "failed to create temporary directory")
+		defer os.RemoveAll(tmp)
+
+		os.WriteFile(filepath.Join(tmp, "main.go"), []byte(`
+package main
+
+func main() {}
+		`), 0o644)
+
+		require.NotPanics(t, func() {
+			_, _, err = goModVersion(tmp)
+		})
+		require.ErrorContains(t, err, "no module information found for package")
+	})
 }
 
 func TestRequiredVersion(t *testing.T) {

--- a/internal/ensure/requiredversion_test.go
+++ b/internal/ensure/requiredversion_test.go
@@ -92,15 +92,15 @@ func TestGoModVersion(t *testing.T) {
 		defer os.RemoveAll(tmp)
 
 		os.WriteFile(filepath.Join(tmp, "main.go"), []byte(`
-package main
+		package main
 
-func main() {}
-		`), 0o644)
+		func main() {}
+				`), 0o644)
 
 		require.NotPanics(t, func() {
 			_, _, err = goModVersion(tmp)
 		})
-		require.ErrorContains(t, err, "no module information found for package")
+		require.ErrorContains(t, err, "go.mod file not found in current directory")
 	})
 }
 

--- a/internal/pin/pin.go
+++ b/internal/pin/pin.go
@@ -28,18 +28,27 @@ const (
 	envValTrue            = "true"
 )
 
-var (
-	requiredVersionError error // Whether the go.mod version check succeeded
-)
-
 // AutoPinOrchestrion automatically runs `pinOrchestrion` if the necessary
 // requirements are not already met. It prints messages to `os.Stderr` to inform
 // the user about what is going on.
 func AutoPinOrchestrion() {
-	if requiredVersionError == nil {
-		// Nothing to do!
+	if os.Getenv(envVarCheckedGoMod) == envValTrue {
+		// A parent process (or ourselves earlier) has already done the check
 		return
 	}
+
+	requiredVersionError := ensure.RequiredVersion()
+	if requiredVersionError == nil {
+		// We're good to go, just make sure we don't do this again
+		_ = os.Setenv(envVarCheckedGoMod, envValTrue)
+		return
+	}
+
+	log.Tracef("Failed to detect required version of orchestrion from go.mod: %v\n", requiredVersionError)
+	if wd, err := os.Getwd(); err == nil {
+		log.Tracef("Working directory: %q\n", wd)
+	}
+	log.Tracef("GOMOD=%s\n", os.Getenv("GOMOD"))
 
 	var (
 		box       = lipgloss.NewStyle()
@@ -87,7 +96,7 @@ func AutoPinOrchestrion() {
 		os.Exit(1)
 	}
 
-	requiredVersionError = nil
+	_ = os.Setenv(envVarCheckedGoMod, envValTrue)
 }
 
 func PinOrchestrion() error {
@@ -138,18 +147,4 @@ func PinOrchestrion() error {
 }
 
 func init() {
-	if os.Getenv(envVarCheckedGoMod) == envValTrue {
-		// A parent process has already done the check for us!!
-		return
-	}
-
-	if requiredVersionError = ensure.RequiredVersion(); requiredVersionError != nil {
-		log.Tracef("Failed to detect required version of orchestrion from go.mod: %v\n", requiredVersionError)
-		if wd, err := os.Getwd(); err == nil {
-			log.Tracef("Working directory: %q\n", wd)
-		}
-		log.Tracef("GOMOD=%s\n", os.Getenv("GOMOD"))
-	} else {
-		_ = os.Setenv(envVarCheckedGoMod, envValTrue)
-	}
 }

--- a/internal/pin/pin_test.go
+++ b/internal/pin/pin_test.go
@@ -6,7 +6,6 @@
 package pin
 
 import (
-	"errors"
 	"fmt"
 	"os"
 	"path/filepath"
@@ -25,9 +24,8 @@ func TestPin(t *testing.T) {
 
 	require.NoError(t, scaffold(tmp))
 	require.NoError(t, os.Chdir(tmp))
-	requiredVersionError = errors.New("test")
 	AutoPinOrchestrion()
-	require.NoError(t, requiredVersionError)
+	require.NotEmpty(t, os.Getenv(envVarCheckedGoMod))
 
 	require.FileExists(t, filepath.Join(tmp, orchestrionToolGo))
 	require.FileExists(t, filepath.Join(tmp, "go.sum"))

